concurrency/DynamicFontDownloadHelper.java

```diff
@@ -0,0 +1,896 @@
+                    AndroidApplicationController.getInstance().showAlert(ErrorState.CONNECTION_ERROR, null);
+                    break;
+                default:
+                    break;
+            }
+        }
+    }
+
+    private void reportFontDownloadOnDemandMetrics(@Nonnull OnDemandFontInfo fontInfo, @Nonnull String event) {
+        String fontTypefaceName = fontInfo.downloadableFont.fontFamily.getTypeFaceName();
+        String fontTimer = getFontDownloadTimer(fontInfo);
+        switch (event) {
+            case ReadingStreamUtil.EVENT_DOWNLOAD_ON_DEMAND_SUCCESS:
+                MetricsManager.getInstance().stopMetricTimerIfExists(ReadingStreamUtil.CONTEXT_DOWNLOAD_ON_DEMAND,
+                        fontTimer, fontTimer);
+                break;
+            case ReadingStreamUtil.EVENT_DOWNLOAD_ON_DEMAND_FAIL:
+                MetricsManager.getInstance().cancelMetricTimer(fontTimer);
+                break;
+            case ReadingStreamUtil.ACTION_DOWNLOAD_ON_DEMAND_CLICK_CANCEL_BUTTON:
+                MetricsManager.getInstance().cancelMetricTimer(fontTimer);
+                break;
+            case ReadingStreamUtil.ACTION_DOWNLOAD_ON_DEMAND_CLICK_DELETE_BUTTON:
+                break;
+            default:
+                return;
+        }
+
+        MetricsManager.getInstance().reportMetric(ReadingStreamUtil.CONTEXT_DOWNLOAD_ON_DEMAND,
+                event + fontTypefaceName);
+        Utils.getFactory().getKindleReaderSDK().getReadingStreamsEncoder().performAction(
+                ReadingStreamUtil.CONTEXT_DOWNLOAD_ON_DEMAND,
+                event + fontTypefaceName);
+    }
+
+    private String getFontDownloadTimer(@Nonnull OnDemandFontInfo fontInfo) {
+        return fontInfo.downloadableFont.fontFamily.getTypeFaceName() + DOWNLOAD_TIMER_SUFFIX;
+    }
+
+    public void cancelDownloadFontOnDemand(IFontInfo font) {
+        OnDemandFontInfo fontInfo = fontInfoMap.get(font.getFontFamily());
+        if (fontInfo != null && fontInfo.isDownloading()) {
+            getAppExpanClient().releaseResourceSet(fontInfo.downloadableFont.resourceSetName);
+            fontInfo.downloadStatus = DownloadStatus.NEED_DOWNLOAD;
+            reportFontDownloadOnDemandMetrics(fontInfo, ReadingStreamUtil.ACTION_DOWNLOAD_ON_DEMAND_CLICK_CANCEL_BUTTON);
+        }
+    }
+
+    /**
+     * Create a symbolic link of the font file to file/, so we don't need to copy files.
+     *
+     * @param f
+     * @return
+     */
+    private boolean createSymlinkForFontFile(File f) {
+        String sourcePath = f.getAbsolutePath();
+        String symLinkPath = Utils.getFactory().getContext().getFilesDir().getAbsolutePath() + "/" + f.getName();
+        File sourceFile = new File(sourcePath);
+
+        if (!sourceFile.exists()) {
+            return false;
+        }
+        try {
+            File symFile = new File(symLinkPath);
+
+            if (symFile.exists()) {
+                symFile.delete();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of deleting a file on this line should be checked and failure should be logged. java.io.File.delete() will simply return false if it fails to delete the file, for example, when the file does not exist. It will not throw unless there is a security exception.